Show the OIDC testing with devservice section first (#44837)
quarkusio/quarkus: docs/src/main/asciidoc/security-oidc-code-flow-authentication.adoc

----

[[code-flow-integration-testing-keycloak-devservices]]
=== Dev Services for Keycloak

Using xref:security-openid-connect-dev-services.adoc[Dev Services for Keycloak] is recommended for integration testing against Keycloak.
`Dev Services for Keycloak` will start and initialize a test container: it will create a `quarkus` realm, a `quarkus-app` client (`secret` secret), and add `alice` (`admin` and `user` roles) and `bob` (`user` role) users, where all of these properties can be customized.

First, prepare `application.properties`.
You can start with a completely empty `application.properties` file as `Dev Services for Keycloak` will register `quarkus.oidc.auth-server-url` pointing to the running test container as well as `quarkus.oidc.client-id=quarkus-app` and `quarkus.oidc.credentials.secret=secret`.

However, if you already have all the required `quarkus-oidc` properties configured, then you only need to associate `quarkus.oidc.auth-server-url` with the `prod` profile for `Dev Services for Keycloak` to start a container.
For example:

[source,properties]
----
%prod.quarkus.oidc.auth-server-url=http://localhost:8180/realms/quarkus
----

> If a custom realm file has to be imported into Keycloak before running the tests, then you can configure `Dev Services for Keycloak` as follows:

Reviewer: ```suggestion
To import a custom realm file into Keycloak before running the tests, configure `Dev services for Keycloak` as shown:
```

[source,properties]
----
%prod.quarkus.oidc.auth-server-url=http://localhost:8180/realms/quarkus
quarkus.keycloak.devservices.realm-path=quarkus-realm.json
----

> Finally, write a test code the same way as it is described in the <<code-flow-integration-testing-wiremock,Wiremock>> section.
> The only difference is that `@QuarkusTestResource` is no longer needed:

Reviewer: ```suggestion
Finally, write the test code as described in the <<code-flow-integration-testing-wiremock,Wiremock>> section.  
The only difference is that `@QuarkusTestResource` is no longer required:
```
